daml-foundations/daml-ghc/daml-prim-src/GHC/Generics.daml

```diff
@@ -0,0 +1,1685 @@
+--       ('C1' ('MetaCons \"IntHash\" 'PrefixI 'False)
+--         ('S1' ('MetaSel 'Nothing
+--                         'NoSourceUnpackedness
+--                         'NoSourceStrictness
+--                         'DecidedLazy)
+--               'UInt'))
+-- @
+--
+-- Currently, only the six unlifted types listed above are generated, but this
+-- may be extended to encompass more unlifted types in the future.
+-- #if 0
+-- *** Limitations
+--
+-- |
+--
+-- /TODO/
+--
+-- /TODO:/ Also clear up confusion about 'Rec0' and 'Rec1' not really indicating recursion.
+--
+-- #endif
+-----------------------------------------------------------------------------
+
+  -- * Generic representation types
+    V1, U1(..), Par1(..), Rec1(..), K1(..), M1(..)
+  , (:+:)(..), (:*:)(..), (:.:)(..)
+
+  -- ** Unboxed representation types
+  --, URec(..)
+  --, type UAddr, type UChar, type UDouble
+  --, type UFloat, type UInt, type UWord
+
+  -- ** Synonyms for convenience
+  , Rec0, R
+  , D1, C1, S1, D, C, S
+
+  -- * Meta-information
+  -- , Datatype(..), Constructor(..), Selector(..)
+  , Fixity(..), FixityI(..), Associativity(..), prec
+  , SourceUnpackedness(..), SourceStrictness(..), DecidedStrictness(..)
+  , Meta(..)
+
+  -- * Generic type classes
+  , Generic(..), Generic1(..)
+
+  ) where
+
+-- We use some base types
+-- import Data.Either ( Either (..) )
+-- import Data.Maybe  ( Maybe(..))
+-- import Data.Ord    ( Down(..) )
+-- import GHC.Integer ( Integer, integerToInt )
+-- import GHC.Prim    ( Addr#, Char#, Double#, Float#, Int#, Word# )
+-- import GHC.Ptr     ( Ptr )
+import GHC.Types
+
+-- Needed for instances
+-- import GHC.Arr     ( Ix )
+-- import GHC.Base    ( Alternative(..), Applicative(..), Functor(..)
+--                   , Monad(..), MonadPlus(..), NonEmpty(..), String, coerce
+--                   , Semigroup(..), Monoid(..) )
+import GHC.Classes ( Eq(..), Ord(..) )
+import GHC.Enum    ( Bounded, Enum )
+-- import GHC.Read    ( Read(..) )
+import GHC.Show    ( Show(..), showString )
+
+-- Needed for metadata
+-- import Data.Proxy   ( Proxy(..) )
+-- import GHC.TypeLits ( KnownSymbol, KnownNat, symbolVal, natVal )
+
+import GHC.CString (fromString)
+import GHC.Integer.Type (fromInteger)
+import Control.Exception.Base
+
+-- | The kind of types with values. For example @Int :: Type@.
+-- [DA] copied from ghc-prim/GHC/Types.hs. Usually set to
+-- type Type = TYPE 'LiftedRep, but we only have one representation.
+type Type = *
+
+ --------------------------------------------------------------------------------
+-- Representation types
+--------------------------------------------------------------------------------
+
+-- | Void: used for datatypes without constructors
+data V1 (p :: k)
+  deriving ( Eq       -- ^ @since 4.9.0.0
+           , Ord      -- ^ @since 4.9.0.0
+--           , Read     -- ^ @since 4.9.0.0
+           , Show     -- ^ @since 4.9.0.0
+--           , Functor  -- ^ @since 4.9.0.0
```

Comment:
Why not Functor?

Reply:
I tried to be as minimal as possible. Deriving functor instances for example is problematic, because the conversion to daml-lf fails for an unhandled coercion (most likely caused by the phantom type in the representation). We might be able to fix this, but I'd rather postpone for later and see if we get the upgradability to work.

Reply:
Can we get a separate ticket for that? Agree not to do it simultaneously, but worth fixing on its own.

Reply:
Done. https://github.com/digital-asset/daml/issues/1344